### DISCUSSION TITLE
Upgrade experience for CMS users

### DISCUSSION THREAD

**Discussion by thibaudcolas** [wagtail/wagtail#12147] [Ideas]
Alongside general [dashboard enhancements in #8325](https://github.com/wagtail/wagtail/discussions/8325), we’ve been looking at ways to improve the "upgrade experience" for CMS users within the CMS. And we could use feedback on what exactly would be most helpful. The big-picture goals are to:

1. Increase CMS users’ awareness of recent and upcoming changes.
2. Increase the likelihood of existing websites keeping up with relevant upgrades.

### Ahead of an upgrade

Currently the main upgrade information comes from the dashboard prompt, which is visible for all admin users, shows for all types of new releases, and links to the release notes:

<img width="653" alt="upgrade notification banner" src="https://github.com/user-attachments/assets/0f333c5a-9697-4962-bef8-ca4148442101">

### After an upgrade

As-is, we badge the “Help” menu and items within (editor guide home, and “New in Wagtail X.Y” editor guide page”, for all users. For Wagtail 3 and 4, we also had a separate dismissible banner for all users:

<img width="510" alt="things in wagtail 4 have changed" src="https://github.com/user-attachments/assets/72903d8f-d6c4-4dff-9459-d2ebbdf10c9f">

## Proposed changes

I’d like to propose introducing more kinds of "upgrade notification" banners. All banners would be:

- Dismissible, so users are more in control of he presence of those notifications.
- Only visible for site users as part of upgrades – not their first time on a site.

If a user is in a scenario where multiple banners could be displayed, we’d display the one that’s highest-priority. Current draft designs:

![current draft designs for upgrade banners](https://github.com/user-attachments/assets/c34f55e2-6d4f-48ca-9e0e-0e9e05ca810d)

### Ahead of an upgrade

Keep the existing "upgrade notification" but split it into three scenarios:

1. Security release available: Banner w/ error styling for admins only (links to dev release notes)
2. Patch release available: Banner w/ warning styling for admins only (links to dev release notes)
3. Feature release available: Banner w/ info styling for admins only – or all users? (links to guide.wagtail.org release page)

Dismissing those banners could either hide them until a new version is _released_, or after a new version has been _upgraded to_ (as-is: not dismissible).

See also: #5541

### After a major or minor upgrade

See above designs, introduce a new "Welcome to Wagtail X.Y" banner for all users. It would be showed for all users who have used the CMS in a previous version (to avoid showing this to first-time users)

### Other scenarios

Here are other scenarios we’ve considered:

- "Version no longer supported" – whether that would warrant being explicitly displayed (see #5541), possibly after a given buffer of time has elapsed after the effective end-of-life date.
- Always-on version number (see [proposed dashboard enhancements](https://github.com/wagtail/wagtail/discussions/8325#discussioncomment-10114586)). So CMS users have a better chance of being aware of what version of Wagtail they should look up documentation for. Currently this information is available but buried within the Settings menu.

## Outside of the CMS

For users who look for extra information about a release, here are proposed changes outside the CMS: 

- **docs.wagtail.org release notes**
    - At the top of the release notes, add a link to guide website page “New in Wagtail X.Y” with screenshots of new features.
    - Add a link to Wagtail release schedule page with end-of-life date, or direct copy of that content for the specific release.
- **guide.wagtail.org “New in Wagtail X.Y” pages**
    - Add a link to announcement blog post for the release
    - Add a call to action to newsletter sign-up to keep up with upcoming releases.
    - Add a link to Wagtail release schedule page with end-of-life date, or direct copy of that content for the release.
- **guide.wagtail.org or wagtail.org**
    - New “Wagtail versions” page with content similar to [https://upgradedjango.com/](https://upgradedjango.com/) / [Release Schedule](https://github.com/wagtail/wagtail/wiki/Release-schedule).
    - Other content relating to Wagtail upgrades? See [Keeping up with upcoming changes in Wagtail](https://wagtail.org/blog/keeping-up-with-upcoming-changes-in-wagtail/).

**Comment by laymonage** [comment]
I like the idea of having the upgrade banner more relevant to the user, i.e. taking into account the last version they were on, whether the version is still in active support, whether there's a new security release, etc.

However, I'd like to know more details on the following:

> "If a user is in a scenario where multiple banners could be displayed, we’d display the one that’s highest-priority."

Are the banners in the screenshot already ordered by priority? If not, could you clarify the ordering?

Also, this would require a few changes in how we check and store the version information:

1. For "Welcome to Wagtail x.y!", we need to remember the Wagtail version that the user last used. This could be stored in the `UserProfile` model, either as a separate field, or in the `dismissibles` JSONField with a string value (or array of numbers). Alternatively, we could use `localStorage`, but it won't be persisted across browsers.
2. For "Wagtail upgrade available", as far as I understand this only requires some styling changes to the current implementation. Although if we implement the other banners, of course we will need to implement the priority ordering as well.
3. For "Security release available" to be accurate, we need to have a machine-readable source of information for each security release or CVE that can tell us which versions were affected.
   - For example: the user is on 6.2.1, there's a 6.2.2 patch release but it also introduced a security issue, which is then fixed in a 6.2.3 release.
   - Displaying "Security release available", while technically correct, the critical nature of the red banner implies it affects the currently installed version – which is not the case, as the issue was introduced in 6.2.2.
   - Not sure how accurate we want this to be – if we don't care about the above case, then this could be as simple as adding a boolean e.g. `"includes_security_fix": true` in [`latest.txt`](https://github.com/wagtail/wagtail/blob/main/scripts/latest.txt) (which, by the way, should really be `.json`...)
   - If we want to handle the above case, then we need to list the versions that are affected (or perhaps the lower and upper bound of the versions that are affected).
4. For "This version of Wagtail (x.y) is no longer supported", we need to either:
   - add a list of all the currently supported versions in the `latest.txt` file, or
   - include the EOL date of the current release to the codebase itself

Alternatively, we could also set up a web service that would respond with the appropriate information for the banner based on a given version. However, this doesn't seem very sustainable compared to hosting a static JSON file like we currently do.

**Reply by thibaudcolas** [reply]
Thank you for the feedback!

Re the priority order, I’ve thought about it and didn’t come up with an order that clearly stood out against other options. My feeling is: 1. "security release" 2. "upgrade available" 3.  "no longer supported"  4. "welcome"

> For "Welcome to Wagtail x.y!", we need to remember the Wagtail version that the user last used.

Yes, I thought I’d ask you whether we should use `dismissibles` or `localStorage`. If the data isn’t there the banner wouldn’t show, so I think the risk is low either way.

> For "Wagtail upgrade available", as far as I understand this only requires some styling changes

Yes.

> For "Security release available"

I don’t think this is meant to be particularly accurate, as it’s very common anyway for there to be security vulnerabilities found in Wagtail that have no impact on a given site. We should adapt the message to make it clear this is about informing the user of the new release. Similar to "upgrade available", but with more immediate attention needed.

> For "This version of Wagtail (x.y) is no longer supported",

My preference would be to find a way to detect this that’s most future-proof.

**Reply by gasman** [reply]
There are a couple of other tricky edge cases with security releases:

* A security fix is released as both 5.2.5 for the LTS branch, and 6.2.2 for the current branch; if the user is already on the LTS release 5.2.5, they should not be told that a newer security release exists
* A user is on 6.2.1; 6.2.2 contains a security fix, and 6.2.3 does not. In this case, the user should be told that a security release exists, even though it's not the latest one.

I think this can be handled by: 1) only considering security releases that are under the same major/minor version as the current one (there should always be one, if it's within support), and 2) rather than a flag `"includes_security_fix": true`, give the version number of the most recent patch release that contains a security fix. The message will then report that a security release is available if that version number is newer than the current one.

**Comment by thibaudcolas** [comment]
## Desired behavior for "Welcome to Wagtail X.Y"

- Only show when the user has previously seen another version of Wagtail (so if I use Wagtail 6.5 and upgrade to 6.6, I will get "Welcome to Wagtail 6.6"). This means the banner won’t be visible in the version we introduce it.

Expected changes:

- Store the whole version number the user has last seen (at login?) (in dismissibles? or elsewhere?)

## Desired behavior for "Upgrade available"

- Shows when a new version of Wagtail is detected (e.g. Wagtail is 6.2, releases endpoint is 6.2.2)
- TBC: whether the dismissing expires at the next-detected patch release, or lasts until the next-detected minor/major release

Expected changes:

- Store the last-dismissed version number
